src/infra/core/src/ingest_service_impl.rs

```diff
@@ -7,44 +7,685 @@
     engine_provisioner: Arc<dyn EngineProvisioner>,
+    object_store_registry: Arc<dyn ObjectStoreRegistry>,
     container_runtime: Arc<ContainerRuntime>,
     run_info_dir: PathBuf,
     cache_dir: PathBuf,
+    time_source: Arc<dyn SystemTimeSource>,
 }
 
+///////////////////////////////////////////////////////////////////////////////
+
 #[component(pub)]
 impl IngestServiceImpl {
     pub fn new(
         dataset_repo: Arc<dyn DatasetRepository>,
         dataset_action_authorizer: Arc<dyn auth::DatasetActionAuthorizer>,
         engine_provisioner: Arc<dyn EngineProvisioner>,
+        object_store_registry: Arc<dyn ObjectStoreRegistry>,
         container_runtime: Arc<ContainerRuntime>,
         run_info_dir: PathBuf,
         cache_dir: PathBuf,
+        time_source: Arc<dyn SystemTimeSource>,
     ) -> Self {
         Self {
             dataset_repo,
             dataset_action_authorizer,
             engine_provisioner,
+            object_store_registry,
             container_runtime,
             run_info_dir,
             cache_dir,
+            time_source,
+        }
+    }
+
+    async fn do_ingest(
+        &self,
+        dataset_ref: &DatasetRef,
+        options: IngestOptions,
+        fetch_override: Option<FetchStep>,
+        get_listener: impl FnOnce(&DatasetHandle) -> Option<Arc<dyn IngestListener>>,
+    ) -> Result<IngestResult, IngestError> {
+        let dataset_handle = self.dataset_repo.resolve_dataset_ref(&dataset_ref).await?;
+
+        self.dataset_action_authorizer
+            .check_action_allowed(&dataset_handle, auth::DatasetAction::Write)
+            .await?;
+
+        let dataset = self
+            .dataset_repo
+            .get_dataset(&dataset_handle.as_local_ref())
+            .await?;
+
+        let listener =
+            get_listener(&dataset_handle).unwrap_or_else(|| Arc::new(NullIngestListener));
+
+        // TODO: Remove this scan after we eliminate Spark ingest
+        let Some((_, polling_source_block)) = dataset
+            .as_metadata_chain()
+            .last_of_type::<SetPollingSource>()
+            .await
+            .int_err()?
+        else {
+            tracing::warn!(
+                %dataset_handle,
+                "Dataset does not define a polling source - considering up-to-date",
+            );
+
+            let result = IngestResult::UpToDate {
+                no_polling_source: true,
+                uncacheable: false,
+            };
+
+            listener.begin();
+            listener.success(&result);
+            return Ok(result);
+        };
+
+        // TODO: Temporarily we continue to default to Spark ingest, but use new engine
+        // if preprocessing query specifies datafusion
+        let engine = polling_source_block
+            .event
+            .preprocess
+            .as_ref()
+            .map(|Transform::Sql(t)| t.engine.as_str());
+
+        if engine == Some("datafusion") {
```

Comment:
This is where Spark and DF ingest code paths branch off

Reply:
Don't we have an enum for engines?

Reply:
Anyone can implement an engine, so we cannot enumerate them all, but I will introduce an enum for those engine we care about (we don't have one yet).